plugins/remind.go

```diff
@@ -51,93 +66,136 @@ func newreminderPlugin(m *seabird.BasicMux, cm *seabird.CommandMux, db *nut.DB)
-
-	// Try to sleep this goroutine until the message needs to be delivered
-	time.Sleep(waitDur)
+func (p *reminderPlugin) joinHandler(b *seabird.Bot, m *irc.Message) {
+	if m.Prefix.Name != b.CurrentNick() {
+		fmt.Println("WE DIDN'T JOIN")
+		return
+	}
 
-	// Send the message
-	b.Send(&irc.Message{
-		Prefix:  &irc.Prefix{},
-		Command: "PRIVMSG",
-		Params:  []string{r.Target, r.Content},
-	})
+	p.roomLock.Lock()
+	defer p.roomLock.Unlock()
+	p.rooms[m.Params[0]] = true
 
-	// Nuke the reminder now that it's been sent
-	err := p.db.Update(func(tx *nut.Tx) error {
-		bucket := tx.Bucket("remind_reminders")
-		return bucket.Delete(r.Key)
-	})
+	p.updateChan <- struct{}{}
+}
 
-	if err != nil {
-		logger.WithError(err).Error("Failed to remove reminder")
+func (p *reminderPlugin) partHandler(b *seabird.Bot, m *irc.Message) {
+	if m.Prefix.Name != b.CurrentNick() {
 		return
 	}
 
-	logger.Info("Dispatched reminder")
+	p.roomLock.Lock()
+	defer p.roomLock.Unlock()
+	delete(p.rooms, m.Params[0])
+
+	p.updateChan <- struct{}{}
 }
 
-func (p *reminderPlugin) dispatchReminders(b *seabird.Bot, tType targetType, target string) {
-	logger := b.GetLogger()
+func (p *reminderPlugin) kickHandler(b *seabird.Bot, m *irc.Message) {
+	if m.Params[1] != b.CurrentNick() {
+		return
+	}
+
+	p.roomLock.Lock()
+	defer p.roomLock.Unlock()
+	delete(p.rooms, m.Params[0])
 
-	reminders := []*reminder{}
+	p.updateChan <- struct{}{}
+}
 
+func (p reminderPlugin) nextReminder() (*reminder, error) {
+	// Find the next reminder we'll have to send
+	var r *reminder
 	err := p.db.View(func(tx *nut.Tx) error {
+		// Grab the room lock for this transaction
+		p.roomLock.Lock()
+		defer p.roomLock.Unlock()
+
 		bucket := tx.Bucket("remind_reminders")
 		cursor := bucket.Cursor()
 
 		v := &reminder{}
 
 		for _, err := cursor.First(v); err == nil; _, err = cursor.Next(v) {
-			if v.TargetType != tType {
-				v = &reminder{}
+			// If it's a channel target and we're not in the room,
+			// we need to skip it
+			if v.TargetType == channelTarget && !p.rooms[v.Target] {
 				continue
 			}
 
-			if target == "" || v.Target != target {
-				v = &reminder{}
+			// If we don't currently have a reminder or the new
+			// reminder should be sent before our current one, we
+			// update it.
+			if r == nil || v.ReminderTime.Before(r.ReminderTime) {
+				r = v
+			}
+		}
+
+		return nil
+	})
+
+	return r, err
+}
+
+func (p *reminderPlugin) remindLoop(b *seabird.Bot) {
+	logger := b.GetLogger()
+
+	for {
+		r, err := p.nextReminder()
+		if err != nil {
+			logger.WithError(err).Error("Transaction failure. Exiting loop.")
+			return
+		}
+
+		var timer <-chan time.Time
+		if r != nil {
+			logger.WithField("reminder", r).Info("Got reminder")
+
+			waitDur := r.ReminderTime.Sub(time.Now())
+			if waitDur <= 0 {
+				p.dispatch(b, r)
 				continue
 			}
 
-			reminders = append(reminders, v)
+			timer = time.After(waitDur)
+		}
 
-			v = &reminder{}
+		select {
+		case <-timer:
```

Comment:
This select will always block. It will either wait for an update to come in (from the updateChan) or for the next reminder to fire (timer).